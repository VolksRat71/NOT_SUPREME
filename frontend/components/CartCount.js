import React, { Component } from 'react';
import { TransitionGroup, CSSTransition } from 'react-transition-group'
import styled from 'styled-components';

const AnimationStyles = styled.span`
    position: relative;
    .count {
        display: block;
        position: relative;
        transition: all 0.4s;
        backface-visibility: hidden;
    }
    // initial state of entered Dot
    .count-enter {
        transform: rotateX(0.5turn);
    }
    .count-enter-active {
        transform: rotateX(0)
    }
    .count-exit {
        top: 0;
        position: absolute;
        transform:rotateX(0)
    }
    .count-exit-active {
        transform: rotateX(0.5turn)
    }
`;

const Dot = styled.div`
    background-color: ${props => props.theme.red};
    color: white;
    border-radius: 50%;
    text-align: center;
    padding: 0.5rem;
    line-height: 2rem;
    min-width: 3rem;
    margin-left: 1rem;
    font-weight: 100;
    font-feature-settings: 'tnum';
    font-variant-numeric: tabular-nums;
`;

const CartCount = ({ count }) => (
    <AnimationStyles>
        <TransitionGroup>
            <CSSTransition
                unmountOnExit
                className="count"
                classNames="count"
                key={count}
                timeout={{
                    enter: 400,
                    exit: 4000
                }}>
                <Dot>{count}</Dot>
            </CSSTransition>
        </TransitionGroup>
    </AnimationStyles>
)

export default CartCount;