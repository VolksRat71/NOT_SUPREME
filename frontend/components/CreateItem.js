import React, { Component } from 'react';
import { Mutation } from 'react-apollo';
import gql from 'graphql-tag';
import Form from './styles/Form';
import formatMoney from '../lib/formatMoney';
import Error from './ErrorMessage';
import Router from 'next/router';

const CREATE_ITEM_MUTATION = gql`
    mutation CREATE_ITEM_MUTATION (
        $title: String!
        $description: String!
        $price: Int!
        $image: String
        $largeImage: String
    ){
        createItem(
            title: $title
            description: $description
            price: $price
            image: $image
            largeImage: $largeImage
        ) {
            id
        }
    }
    `

class CreateItem extends Component {
    state = {
        title: 'Mask Cat',
        description: 'Judgy Bitch will start at you.',
        image: 'https://image.shutterstock.com/image-photo/cat-medical-mask-protective-antiviral-260nw-1684423789.jpg',
        largeImage: 'https://image.shutterstock.com/image-photo/cat-medical-mask-protective-antiviral-260nw-1684423789.jpg',
        price: 20
    }

    handleChange = e => {
        const { name, type, value } = e.target;
        const val = type === 'number' ? parseFloat(value) : value
        this.setState({ [name]: val })
    }

    uploadFile = async e => {
        console.log("uploading file ...");
        const files = e.target.files;
        const data = new FormData();
        data.append('file',  files[0]);
        data.append('upload_present', 'NotSupreme');

        const res = await fetch(
            'https://api.cloudinary.com/v1_1/notsupreme/image/upload',
            {
                method: 'POST',
                body: data
            }
        )

        const file = await res.json();
        this.setState({
            image: file.secure_url,
            largeImage: file.eager[0].secure_url
        });
    }

    render() {
        return (
            <Mutation
                mutation={CREATE_ITEM_MUTATION}
                variables={this.state}>
                {(createItem, { loading, error }) => (
                    <Form onSubmit={async e => {
                        e.preventDefault();
                        const res = await createItem();
                        Router.push({
                            pathname: '/item',
                            query: { id: res.data.createItem.id }
                        })
                    }}>
                        <Error error={error} />
                        <fieldset disabled={loading} aria-busy={loading}>
                            <label htmlFor="file">
                                Image
                                <input
                                    type="file"
                                    id="file"
                                    name="file"
                                    placeholder="Upload an image"
                                    required
                                    onChange={this.uploadFile}
                                />
                            </label>
                            <label htmlFor="title">
                                Title
                                <input
                                    type="text"
                                    id="title"
                                    name="title"
                                    placeholder="Title"
                                    required
                                    value={this.state.title}
                                    onChange={this.handleChange}
                                />
                            </label>
                            <label htmlFor="price">
                                Price
                                <input
                                    type="number"
                                    id="price"
                                    name="price"
                                    placeholder="Price"
                                    required
                                    value={this.state.price}
                                    onChange={this.handleChange}
                                />
                            </label>
                            <label htmlFor="description">
                                Description
                                <textarea
                                    id="description"
                                    name="description"
                                    placeholder="Enter A Description"
                                    required
                                    value={this.state.description}
                                    onChange={this.handleChange}
                                />
                            </label>
                            <button
                                type="submit"
                            >Submit</button>
                        </fieldset>
                    </Form>
                )}
            </Mutation>
        )
    }
}

export default CreateItem;
export { CREATE_ITEM_MUTATION };