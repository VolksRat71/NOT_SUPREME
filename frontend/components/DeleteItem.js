import React, { Component } from 'react';
import { Mutation } from 'react-apollo';
import gql from 'graphql-tag';
import { ALL_ITEMS_QUERY } from './Items';

const DELETE_ITEM_MUTATION = gql`
    mutation DELETE_ITEM_MUTATION($id: ID!) {
        deleteItem(id: $id){
            id
        }
    }
`

class DeleteItem extends Component {
    update = (cache, payload) => {
        // manually update cache so it matches server
        const data = cache.readQuery({query: ALL_ITEMS_QUERY});
        console.log(data);
        data.items = data.items.filter(item => item.id !== payload.data.deleteItem.id)
        cache.writeQuery({query: ALL_ITEMS_QUERY, data})
    }
    render() {
        return (
            <Mutation
            mutation={ DELETE_ITEM_MUTATION }
            variables={{ id: this.props.id }}
            update={this.update}>
                {(deleteItem, { error }) => (
                <button onClick={() => {
                if(confirm('Are you sure you want to delete this item?')) {
                    deleteItem();
                }}}>
                { this.props.children }
                </button>)}
            </Mutation>
        )
    }
}

export default DeleteItem;