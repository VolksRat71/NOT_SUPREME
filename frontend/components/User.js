import { Query } from 'react-apollo';
import gql from 'graphql-tag';
import propTypes from 'prop-types';

const CURRENT_USER_QUERY = gql`
    query {
        me {
            id
            email
            name
            permissions
        }
    }
`

const User = props => (
    <Query {...props} query={CURRENT_USER_QUERY}>
        {payload => props.children(payload)}
    </Query>
);

User.propTypes = {
    children: propTypes.func.isRequired
}

export default User;
export { CURRENT_USER_QUERY };