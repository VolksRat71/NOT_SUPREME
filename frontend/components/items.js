import React, { Component } from 'react';
import { Query } from 'react-apollo';
import gql from 'graphql-tag';

const ALL_ITEMS_QUERY = gql`
    query ALL_ITEMS_QUERY {
        items {
            id
            title
            price
            description
            image
            largeImage
        }
    }
`

class Items extends Component {
    render() {
        return (
            <div>
                <p>Items!</p>
                <Query query={ALL_ITEMS_QUERY}>
                    {({ data, error, loading }) => {
                        if (loading) return <p>Loading...</p>
                        if (error) return <p>Error: {error.message}</p>
                        return <p>Items!</p>
                    }}
                </Query>
            </div>
        );
    };
};

export default Items;