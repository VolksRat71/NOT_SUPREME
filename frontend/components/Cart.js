import React, { Component } from 'react';
import { Query, Mutation } from 'react-apollo';
import gql from 'graphql-tag';
import { adopt } from 'react-adopt';
import CartStyles from './styles/CartStyles';
import Supreme from './styles/Supreme';
import CloseButton from './styles/CloseButton';
import SickButton from './styles/SickButton';
import User from './User';
import CartItem from './CartItem';
import calcTotalPrice from '../lib/calcTotalPrice';
import formatMoney from '../lib/formatMoney';
import takeMyMoney from './TakeMyMoney';
import TakeMyMoney from './TakeMyMoney';

const LOCAL_STATE_QUERY = gql`
    query {
        cartOpen @client
    }
`;

const TOGGLE_CART_MUTATION = gql`
    mutation {
        toggleCart @client
    }
`;

const Composed = adopt({
    user: ({ render }) => <User>{render}</User>,
    toggleCart: ({ render }) => <Mutation mutation={TOGGLE_CART_MUTATION}>{render}</Mutation>,
    localState: ({ render }) => <Query query={LOCAL_STATE_QUERY}>{render}</Query>
});

const Cart = () => (
    <Composed>
        {({ user, toggleCart, localState }) => {
            const me = user.data.me;
            if (!me) return null;
            const firstName = me.name.split(" ")[0]
            return (
                <CartStyles open={localState.data.cartOpen}>
                    <header>
                        <CloseButton onClick={toggleCart} title="close">&times;</CloseButton>
                        <Supreme>{firstName.toUpperCase()}'S CART</Supreme>
                        <p>You have {me.cart.length} item{me.cart.length === 1 ? '' : 's'} in your cart.</p>
                    </header>
                    <ul>
                        {me.cart.map(cartItem => <CartItem key={cartItem.id} cartItem={cartItem} />)}
                    </ul>
                    <footer>
                        {me.cart.length &&
                            <>
                                <p>{formatMoney(calcTotalPrice(me.cart))}</p>
                                <TakeMyMoney>
                                    <SickButton>Checkout</SickButton>
                                </TakeMyMoney>
                            </>
                        }
                    </footer>
                </CartStyles>
            )
        }}
    </Composed>
)

export default Cart;
export { LOCAL_STATE_QUERY, TOGGLE_CART_MUTATION };