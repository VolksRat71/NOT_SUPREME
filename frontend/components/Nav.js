import Link from 'next/link';
import { Mutation } from 'react-apollo';
import NavStyles from './styles/NavStyles';
import User from './User';
import SignOut from './SignOut';
import { TOGGLE_CART_MUTATION } from './Cart';

const Nav = props => (
    <User>
        {({ data: { me } }) => (
            <NavStyles>
                <Link href="/items">
                    <a>
                        <p>Shop</p>
                    </a>
                </Link>
                {me && (
                    <>
                        <Link href="/sell">
                            <a>
                                <p>Sell</p>
                            </a>
                        </Link>
                        <Link href="/orders">
                            <a>
                                <p>Orders</p>
                            </a>
                        </Link>
                        <Link href="/me">
                            <a>
                                <p>Account</p>
                            </a>
                        </Link>
                        <Mutation mutation={TOGGLE_CART_MUTATION}>
                            {toggleCart => (
                                <a onClick={toggleCart}>My Cart</a>
                            )}
                        </Mutation>
                        <SignOut />
                    </>
                )}
                {!me && (
                    <Link href="/signup">
                        <a>
                            <p>Sign In</p>
                        </a>
                    </Link>
                )}
            </NavStyles>
        )}
    </User>
)

export default Nav;