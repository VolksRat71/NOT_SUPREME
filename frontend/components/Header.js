import Nav from './Nav'

const Header = props => (
    <>
        <div className="bar">
            <a href="/">Not Supreme</a>
        </div>
        <div className="sub-bar">
            <p>Search</p>
        </div>
        <div>Cart</div>
        <Nav />
    </>
)

export default Header;