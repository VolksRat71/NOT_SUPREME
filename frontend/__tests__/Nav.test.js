import { mount } from 'enzyme';
import wait from 'waait';
import toJSON from 'enzyme-to-json';
import { MockedProvider } from 'react-apollo/test-utils';
import { fakeUser, fakeCartItem } from '../lib/testUtils';
import Nav from '../components/Nav';
import { CURRENT_USER_QUERY } from '../components/User';

const notSignedInMocks = [
    {
        request: { query: CURRENT_USER_QUERY },
        result: { data: { me: null } },
    },
];

const signedInMocks = [
    {
        request: { query: CURRENT_USER_QUERY },
        result: {
            data: {
                me: {
                    ...fakeUser(),
                    cart: [],
                },
            },
        },
    },
];


const signedInMocksWithCartItems = [
    {
        request: { query: CURRENT_USER_QUERY },
        result: {
            data: {
                me: {
                    ...fakeUser(),
                    cart: [fakeCartItem(), fakeCartItem(), fakeCartItem()],
                },
            },
        },
    },
];

describe('<Nav />', () => {
    test('Renders a minimal <Nav /> when signed out ', async () => {
        const wrapper = mount(
            <MockedProvider mocks={notSignedInMocks}>
                <Nav />
            </MockedProvider>
        )
        await wait();
        const nav = wrapper.find('[data-test="nav"]')
        expect(toJSON(nav)).toMatchSnapshot()
    })

    test('Renders a full <Nav /> when signed in ', async () => {
        const wrapper = mount(
            <MockedProvider mocks={signedInMocks}>
                <Nav />
            </MockedProvider>
        )
        await wait();
        const nav = wrapper.find('[data-test="nav"]');
        expect(toJSON(nav)).toMatchSnapshot();
    })

    test('Renders the amount of items in the cart', async () => {
        const wrapper = mount(
            <MockedProvider mocks={signedInMocksWithCartItems}>
                <Nav />
            </MockedProvider>
        );
        await wait();
        wrapper.update();
        const nav = wrapper.find('[data-test="nav"]');
        const count = nav.find('div.count');
        expect(toJSON(count)).toMatchSnapshot()
    });
})
