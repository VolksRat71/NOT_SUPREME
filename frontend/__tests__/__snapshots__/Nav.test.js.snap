// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Nav /> Renders a full <Nav /> when signed in  1`] = `
Array [
  <NavStyles
    data-test="nav"
  >
    <ul
      className="NavStyles-sc-1hbmsij-0 esuWmp"
      data-test="nav"
    >
      <Link
        href="/items"
      >
        <a
          href="/items"
          onClick={[Function]}
        >
          <p>
            Shop
          </p>
        </a>
      </Link>
      <Link
        href="/signup"
      >
        <a
          href="/signup"
          onClick={[Function]}
        >
          <p>
            Sign In
          </p>
        </a>
      </Link>
    </ul>
  </NavStyles>,
  <ul
    className="NavStyles-sc-1hbmsij-0 esuWmp"
    data-test="nav"
  >
    <Link
      href="/items"
    >
      <a
        href="/items"
        onClick={[Function]}
      >
        <p>
          Shop
        </p>
      </a>
    </Link>
    <Link
      href="/signup"
    >
      <a
        href="/signup"
        onClick={[Function]}
      >
        <p>
          Sign In
        </p>
      </a>
    </Link>
  </ul>,
]
`;

exports[`<Nav /> Renders a minimal <Nav /> when signed out  1`] = `
Array [
  <NavStyles
    data-test="nav"
  >
    <ul
      className="NavStyles-sc-1hbmsij-0 esuWmp"
      data-test="nav"
    >
      <Link
        href="/items"
      >
        <a
          href="/items"
          onClick={[Function]}
        >
          <p>
            Shop
          </p>
        </a>
      </Link>
      <Link
        href="/signup"
      >
        <a
          href="/signup"
          onClick={[Function]}
        >
          <p>
            Sign In
          </p>
        </a>
      </Link>
    </ul>
  </NavStyles>,
  <ul
    className="NavStyles-sc-1hbmsij-0 esuWmp"
    data-test="nav"
  >
    <Link
      href="/items"
    >
      <a
        href="/items"
        onClick={[Function]}
      >
        <p>
          Shop
        </p>
      </a>
    </Link>
    <Link
      href="/signup"
    >
      <a
        href="/signup"
        onClick={[Function]}
      >
        <p>
          Sign In
        </p>
      </a>
    </Link>
  </ul>,
]
`;

exports[`<Nav /> Renders the amount of items in the cart 1`] = `
<div
  className="count CartCount__Dot-sc-1lmkjzl-1 gfVdLM"
>
  9
</div>
`;
